packages/realm_dart/lib/src/logging.dart

```diff
@@ -0,0 +1,178 @@
+  late final server = _LeafLogCategory('Server', this);
+}
+
+class _ClientLogCategory extends RealmLogCategory {
+  _ClientLogCategory(RealmLogCategory parent) : super._('Client', parent);
+
+  late final changeset = _LeafLogCategory('Changeset', this);
+  late final network = _LeafLogCategory('Network', this);
+  late final reset = _LeafLogCategory('Reset', this);
+  late final session = _LeafLogCategory('Session', this);
+}
+
+class _StorageLogCategory extends RealmLogCategory {
+  _StorageLogCategory(RealmLogCategory parent) : super._('Storage', parent);
+
+  late final notification = _LeafLogCategory('Notification', this);
+  late final object = _LeafLogCategory('Object', this);
+  late final query = _LeafLogCategory('Query', this);
+  late final transaction = _LeafLogCategory('Transaction', this);
+}
+
+/// Specifies the criticality level above which messages will be logged
+/// by the default sync client logger.
+/// {@category Realm}
+enum RealmLogLevel {
+  /// Log everything. This will seriously harm the performance of the
+  /// sync client and should never be used in production scenarios.
+  all(Level.ALL),
+
+  /// A version of [debug] that allows for very high volume output.
+  /// This may seriously affect the performance of the sync client.
+  trace(Level.FINEST),
+
+  /// Reveal information that can aid debugging, no longer paying
+  /// attention to efficiency.
+  debug(Level.FINER),
+
+  /// Same as [info], but prioritize completeness over minimalism.
+  detail(Level.FINE),
+
+  /// Log operational sync client messages, but in a minimalist fashion to
+  /// avoid general overhead from logging and to keep volume down.
+  info(Level.INFO),
+
+  /// Log errors and warnings.
+  warn(Level.WARNING),
+
+  /// Log errors only.
+  error(Level.SEVERE),
+
+  /// Log only fatal errors.
+  fatal(Level.SHOUT),
+
+  /// Turn off logging.
+  off(Level.OFF),
+  ;
+
+  /// The [Level] from package [logging](https://pub.dev/packages/logging) that
+  /// corresponds to this [RealmLogLevel].
+  final Level level;
+
+  const RealmLogLevel(this.level);
+}
+
+/// A record of a log message from the Realm SDK.
+typedef RealmLogRecord = ({RealmLogCategory category, RealmLogLevel level, String message});
+
+/// A logger that logs messages from the Realm SDK.
+class RealmLogger {
+  static final _controller = StreamController<RealmLogRecord>.broadcast(
+    onListen: () => realmCore.loggerAttach(),
+    onCancel: () => realmCore.loggerDetach(),
+  );
+
+  const RealmLogger();
+
+  /// Set the log [level] for the given [category].
+  void setLogLevel(RealmLogLevel level, {RealmLogCategory? category}) {
+    category ??= RealmLogCategory.realm;
+    realmCore.setLogLevel(level, category: category);
+  }
+
+  Stream<RealmLogRecord> get onRecord => _controller.stream;
+
+  void _raise(RealmLogRecord record) {
+    _controller.add(record);
+  }
+
+  void _log(RealmLogLevel level, Object message, {RealmLogCategory? category}) {
```

Comment:
In what other types would we expect the `message` to be passed as other than `String`?

Reply:
Dart's default logger package allows objects to be passed for message. This just follows that convention. Previously I considered keeping log method public, since that would be less of a breaking change. However, I decided against that.

Reply:
Ah okay, thanks 👍